### PR TITLE
Deprecate awsKmsKeyArn for provider and functoin kmsKeyArn

### PR DESCRIPTION
**Pull request by fredericbarthelet**
Addresses: #8261

### CODE REVIEW

File: lib/plugins/aws/package/compile/functions/index.js [hunk 1]
```diff
@@ -210,12 +210,27 @@ class AwsCompileFunctions {
...
       const serviceObj = this.serverless.service.serviceObject;
-      if (functionObject.awsKmsKeyArn || (serviceObj && serviceObj.awsKmsKeyArn)) {
-        const arn = functionObject.awsKmsKeyArn || (serviceObj && serviceObj.awsKmsKeyArn);
+      if (functionObject.kmsKeyArn || this.serverless.service.provider.kmsKeyArn) {
+        kmsKeyArn = functionObject.kmsKeyArn || this.serverless.service.provider.kmsKeyArn;
+      } else if (functionObject.awsKmsKeyArn) {
+        this.serverless._logDeprecation(
+          'FUNCTION_DEPRECATED_PROPERTY_AWS_KMS_KEY_ARN',
+          'Starting with next major version, awsKmsKeyArn function property will be replaced by kmsKeyArn.'
+        );
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/functions/index.js:221]: It'll be nice to show this deprecation _unconditionally_ when `functionObject.awsKmsKeyArn` is set

(currently we show it only if it's set and there's no `functionObject.kmsKeyArn` or `this.serverless.service.provider.kmsKeyArn`, which introduces a possibility of throwing config validation error with next major without any deprecation notice in this major)

File: lib/plugins/aws/package/compile/functions/index.js [hunk 2]
```diff
@@ -210,12 +210,27 @@ class AwsCompileFunctions {
...
+        kmsKeyArn = functionObject.awsKmsKeyArn;
+      } else if (serviceObj && serviceObj.awsKmsKeyArn) {
+        this.serverless._logDeprecation(
+          'SERVICE_DEPRECATED_PROPERTY_AWS_KMS_KEY_ARN',
+          'Starting with next major version, awsKmsKeyArn service property will be replaced by provider.kmsKeyArn.'
+        );
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/functions/index.js:227]: Same here

File: lib/plugins/aws/package/compile/functions/index.js [hunk 3]
```diff
@@ -210,12 +210,27 @@ class AwsCompileFunctions {
...
-        const arn = functionObject.awsKmsKeyArn || (serviceObj && serviceObj.awsKmsKeyArn);
+      if (functionObject.kmsKeyArn || this.serverless.service.provider.kmsKeyArn) {
+        kmsKeyArn = functionObject.kmsKeyArn || this.serverless.service.provider.kmsKeyArn;
+      } else if (functionObject.awsKmsKeyArn) {
+        this.serverless._logDeprecation(
+          'FUNCTION_DEPRECATED_PROPERTY_AWS_KMS_KEY_ARN',
+          'Starting with next major version, awsKmsKeyArn function property will be replaced by kmsKeyArn.'
+        );
+        kmsKeyArn = functionObject.awsKmsKeyArn;
+      } else if (serviceObj && serviceObj.awsKmsKeyArn) {
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/functions/index.js:223]: `serviceObj` is guaranteed to be here (although it's possible it's written that way as tests work on incomplete instances).

Anyway point is, that if relying on `serviceObj.awsKmsKeyArn` unconditionally won't break the existing tests, let's not check for `serviceObj` existence

File: lib/plugins/aws/package/compile/functions/index.js [hunk 4]
```diff
@@ -210,12 +210,27 @@ class AwsCompileFunctions {
...
+      let kmsKeyArn;
       const serviceObj = this.serverless.service.serviceObject;
-      if (functionObject.awsKmsKeyArn || (serviceObj && serviceObj.awsKmsKeyArn)) {
-        const arn = functionObject.awsKmsKeyArn || (serviceObj && serviceObj.awsKmsKeyArn);
+      if (functionObject.kmsKeyArn || this.serverless.service.provider.kmsKeyArn) {
+        kmsKeyArn = functionObject.kmsKeyArn || this.serverless.service.provider.kmsKeyArn;
+      } else if (functionObject.awsKmsKeyArn) {
+        this.serverless._logDeprecation(
+          'FUNCTION_DEPRECATED_PROPERTY_AWS_KMS_KEY_ARN',
+          'Starting with next major version, awsKmsKeyArn function property will be replaced by kmsKeyArn.'
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/functions/index.js:220]: Let's remove trailing `.`

File: lib/plugins/aws/package/compile/functions/index.js [hunk 5]
```diff
@@ -210,12 +210,27 @@ class AwsCompileFunctions {
...
+        );
+        kmsKeyArn = functionObject.awsKmsKeyArn;
+      } else if (serviceObj && serviceObj.awsKmsKeyArn) {
+        this.serverless._logDeprecation(
+          'SERVICE_DEPRECATED_PROPERTY_AWS_KMS_KEY_ARN',
+          'Starting with next major version, awsKmsKeyArn service property will be replaced by provider.kmsKeyArn.'
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/functions/index.js:226]: Let's remove trailing `.`

File: lib/configSchema.js [hunk 6]
```diff
@@ -125,7 +125,7 @@ const schema = {
   additionalProperties: false,
   required: ['provider', 'service'],
   definitions: {
-    // TODO: awsKmsArn definition to be moved to lib/plugins/aws/provider/awsProvider.js once service.awsKmsKeyArn moved to provider.awsKmsKeyArn, see https://github.com/serverless/serverless/issues/8261
+    // TODO: awsKmsArn definition to be moved to lib/plugins/aws/provider/awsProvider.js once service.awsKmsKeyArn removed with v2.0.0, see https://github.com/serverless/serverless/issues/8261
```
Comment (medikoo) [inline lib/configSchema.js:128]: _is removed with v3.0.0_

**Review by medikoo — CHANGES_REQUESTED** [review]
Thank you @fredericbarthelet, that looks very good! I've picked just on a few minor, mostly style related things

File: lib/plugins/aws/package/compile/functions/index.js (hunk 3 above)
Comment (fredericbarthelet) [reply lib/plugins/aws/package/compile/functions/index.js:223]: 57 tests do fail if serviceObj is not checked for existence.

**Comment by fredericbarthelet** [comment]
@medikoo I updated my PR with your comments. Let me know what you think :)

File: lib/plugins/aws/package/compile/functions/index.js (hunk 3 above)
Comment (medikoo) [reply lib/plugins/aws/package/compile/functions/index.js]: Ok, thanks for checking, let's leave it then (tests require refactor to `runServerless` but ofc that's not in scope here)

File: lib/plugins/aws/package/compile/functions/index.js [hunk 7]
```diff
@@ -210,12 +210,32 @@ class AwsCompileFunctions {
         }
       }
 
+      let kmsKeyArn;
       const serviceObj = this.serverless.service.serviceObject;
-      if (functionObject.awsKmsKeyArn || (serviceObj && serviceObj.awsKmsKeyArn)) {
-        const arn = functionObject.awsKmsKeyArn || (serviceObj && serviceObj.awsKmsKeyArn);
+      if (serviceObj && serviceObj.awsKmsKeyArn) {
+        this.serverless._logDeprecation(
+          'SERVICE_DEPRECATED_PROPERTY_AWS_KMS_KEY_ARN',
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/functions/index.js:217]: Let's not use `DEPRECATED` in deprecation codes (seems verbose, when it's clear we deal with deprecation code)

Also I see we use two different deprecation codes, while we document it under one code.

I think it's ok to use same code for both (which could be: `AWS_KMS_KEY_ARN`), and it's ok that messages are different (however if both will be triggered only first will be shown, still that seems acceptable).

What's important, is that code is used in generated link, so when introducing not documented deprecation code, invalid link is generated

**Review by medikoo — CHANGES_REQUESTED** [review]
Thank you @fredericbarthelet, looks great. Sorry, spotted one more issue related to deprecation codes. Let me know

**Comment by fredericbarthelet** [comment]
@medikoo updated accordingly :)

File: docs/deprecations.md [hunk 8]
```diff
@@ -73,3 +73,9 @@ Please use `onUnauthenticatedRequest` instead. `allowUnauthenticated` will be re
 ## `bin/serverless`
 
 Please use `bin/serverless.js` instead. `bin/serverless` will be removed with v2.0.0
+
+<a name="AWS_KMS_KEY_ARN"><div>&nbsp;</div></a>
+
+## awsKmsKeyArn references
+
+Plase use `provider.kmsKeyArn` and `functions[].kmsKeyArn`. `service.awsKmsKeyArn` and `functions[].awsKmsKeyArn` will be removed with v2.0.0
```
Comment (medikoo) [inline docs/deprecations.md:81]: _will be removed with v3.0.0_ :)

**Review by medikoo — COMMENTED** [review]
Thank you @fredericbarthelet ! It all looks great, just version number is not right in deprecation doc :)

File: docs/deprecations.md (hunk 8 above)
Comment (fredericbarthelet) [reply docs/deprecations.md]: Fixed :)

**Review by medikoo — APPROVED** [review]
Thank you @fredericbarthelet !

**Pull request merged**